src/core/src/core_logic/PatchInstaller.py

```diff
@@ -282,7 +349,202 @@ def install_updates(self, maintenance_window, package_manager, simulate=False):
+        maintenance_window_batch_cutoff_reached = False
+
+        # remaining_packages are the packages which are not attempted to install due to there is not enough remaining time in maintenance window to install packages in batches.
+        # These packages will be attempted in sequential installation if there is enough time in maintenance window to install package sequentially.
+        remaining_packages = []
+        remaining_package_versions = []
+        
+        # failed_packages are the packages which are failed to install in batch patching. These packages will be attempted again in sequential patching if there is 
+        # enough time remaining in maintenance window.
+        failed_packages = []
+        failed_package_versions = []
+
+        for batch_index in range(0, number_of_batches):
+            per_batch_installation_stopwatch = Stopwatch(self.env_layer, self.telemetry_writer, self.composite_logger)
+            per_batch_installation_stopwatch.start()
+
+            # Extension state check
+            if self.lifecycle_manager is not None:
+                self.lifecycle_manager.lifecycle_status_check()
+
+            begin_index = batch_index * Constants.MAX_BATCH_SIZE_FOR_PACKAGES
+            end_index = begin_index + Constants.MAX_BATCH_SIZE_FOR_PACKAGES - 1
+            end_index = min(end_index, len(packages) - 1)
+
+            packages_in_batch = []
+            package_versions_in_batch = []
+            skip_packages = []
+            already_installed_packages = []
+
+            for index in range(begin_index, end_index + 1):
+                if package_versions[index] == Constants.UA_ESM_REQUIRED:
+                    skip_packages.append(packages[index])
+                    self.status_handler.set_package_install_status(package_manager.get_product_name(packages[index]), str(package_versions[index]), Constants.NOT_SELECTED)
+                elif packages[index] not in self.last_still_needed_packages:
+                    # Could have got installed as dependent package of some other package. Package installation status could also have been set.
+                    already_installed_packages.append(packages[index])
+                    self.attempted_parent_package_install_count += 1
+                    self.successful_parent_package_install_count += 1
+                else:
+                    packages_in_batch.append(packages[index])
+                    package_versions_in_batch.append(package_versions[index])
+
+            if len(already_installed_packages) > 0:
+                self.composite_logger.log("Following packages are already installed. Could have got installed as dependent package of some other package " + str(already_installed_packages))
+
+            if len(skip_packages) > 0:
+                self.composite_logger.log("[Skipping packages " + str(skip_packages) + " - requires Ubuntu Advantage for Infrastructure with Extended Security Maintenance]")
```

Comment:
note, separate from this PR. Needs to be reviewed long term for the parent PatchInstaller to be agnostic. @rane-rajasi